config/tools/config_generator.cpp

```diff
@@ -0,0 +1,935 @@
+                      << ", in Atomizer configuration generation."
+                      << std::endl;
+        }
+    }
+
+    void config_generator::load_template(
+        const std::string filename,
+        std::map<std::string, std::string>& config_map) {
+        std::ifstream file(filename);
+        assert(file.good());
+        std::string line;
+        while(std::getline(file, line)) {
+            std::istringstream line_stream(line);
+            std::string key;
+            if(std::getline(line_stream, key, '=')) {
+                std::string value;
+                if(std::getline(line_stream, value)) {
+                    config_map.emplace(key, value);
+                }
+            }
+        }
+    }
+
+    void config_generator::write_generated_config_to_file(
+        const std::string& _config_file) {
+        std::ofstream outFile;
+        outFile.open(_config_file);
+        outFile << m_new_config.str();
+        outFile.close();
+    }
+
+    [[nodiscard]] auto
+    config_generator::copy_to_build_dir(const std::string filename) -> bool {
+        std::filesystem::path cwd = std::filesystem::current_path();
+        cwd.append(filename);
+        if(std::filesystem::exists(filename)) {
+            const auto copyOptions
+                = std::filesystem::copy_options::overwrite_existing;
+            // Copy and remove file if we are not in build currently
+            if(std::filesystem::current_path() != m_build_dir) {
+                std::filesystem::copy(cwd, m_build_dir, copyOptions);
+                std::filesystem::remove(cwd);
+            }
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    void config_generator::copy_templates_to_build_dir() {
+        std::filesystem::path config_dir = m_project_root_dir;
+        config_dir.append("config").append("tools");
+        std::filesystem::path build_config_dir = m_build_dir;
+        build_config_dir.append("config").append("tools");
+        for(auto const& dir_entry :
+            std::filesystem::directory_iterator{config_dir}) {
+            std::string filename = dir_entry.path().filename();
+            std::string match_str = ".tmpl";
+            std::string tmp_str = filename.substr(filename.size() - 5, 5);
+            if(tmp_str == match_str) {
+                const auto copyOptions
+                    = std::filesystem::copy_options::overwrite_existing;
+                std::filesystem::copy(dir_entry,
+                                      build_config_dir,
+                                      copyOptions);
+                std::cout << "Copying " << dir_entry.path().string() << " to "
+                          << build_config_dir.string() << std::endl;
+            }
+        }
+    }
+
+    template<typename T>
+    [[maybe_unused]] auto config_generator::find_value(const std::string key,
+                                                       T& output) -> bool {
```

Comment:
Why do you prefer an out-parameter rather than a `std::optional<T>` return value?

Reply:
std::optional is a new concept for me with this project. Will change it the way i think you are suggesting though. However, i suppose, i am not sure what i would return in the instances where it was not found.

Reply:
No worries.

The benefit of `optional` is that it allows you to signal the potential absence of a value. In this case, then, using `optional` lets you signal the equivalent of the current `return false` cases (e.g., unknown template parameter type).

If we wanted to be more type-safe/efficiently-serializable (though, the benefits of doing so in this case might be negligible), we might define an error-status type for this function (or related functions), and return a `std::variant<T, error_status_type>` to let us specifically report what went wrong (allowing much more efficient serialization and dispatch); that would offer something a little closer to what exceptions might offer while still avoiding a lot of the drawbacks/overhead. We do something similar in the validation code so that we can serialize only an error code in messages rather than a full string which has to be parsed on the other side.

For this case though, just using an `optional` is probably enough. Hope that's helpful context! If you need more clarity on my question (note: I'm not necessarily against the out-param, though usually my default is to avoid side-effects), just let me know.

Reply:
Ended up figuring this out i believe.

DONE